Python migration of files for backup process to AWS S3

kimlew opened the pull request:
@kcharter 
This migration from Python 2.7 to 3.6 of files for the process of backup to AWS S3 includes:
- changes from running the 2to3 tool
- some manual changes

For setup and testing, it uses:
- a Dockerfile - for the correct environment setup
- a Makefile and test_tablesnap.sh - to test that the changed files work

The affected files are:
- tablesnap
- tableslurp
- tablechop
- tablesnap-sunset
- tablesnap-sync-manifest

kimlew committed:
Ignores 2 Python environment dirs & *.bak

kimlew committed:
Adds instructions to build & run container with tablesnap requirements & Python envs

kimlew committed:
Adds instructions for image based on ubuntu 18.04

kimlew committed:
Modifies tablesnap file with 2to3 changes plus manual changes

kimlew committed:
Modifies tableslurp file with 2to3 changes plus manual changes

kimlew committed:
Changes test_py2.7 & test_py3.6 to test tableslurp & use -e USER=root

kimlew committed:
Modifies tablesnap-sync-manifest with 2to3 changes plus manual changes

kimlew committed:
Modifies tablesnap-sunset with 2to3 changes plus manual changes

kimlew committed:
Adds lines to test_py2.7 & test_py3.6 to test other files like tablechop

kimlew committed:
Replaces tabs with spaces based on PyCharm hints & spacing in less

kimlew committed:
Modifies tablechop with 2to3 changes

kimlew committed:
Adds RUN commands to mkdir /tmp/tablesnap-test & installs aws cli, curl, zip

kimlew committed:
Adds test for tablesnap with passed params from test_tablesnap.sh

kimlew committed:
Initial commit of script to work with Makefile's tablesnap test

kimlew committed:
Adds .idea to ignore when opened in PyCharm

kimlew committed:
Adds - after << so HEREdoc can be indented

kimlew committed:
Adds 2to3 changes: list(). But change back Except to: from Queue import Queue

kimlew committed:
Adds 2nd longer timeout so mv has more time to copy files to S3

kimlew committed:
Adds 2nd longer timeout & adds missed BUCKET & HARDLINKS_PATH to usage()

kimlew committed:
Adds 2to3 change: list(). But change back Except to: from Queue import Queue

kimlew committed:
Adds 2to3 change: list()

kcharter commented on tablechop:
```diff
@@ -75,25 +75,27 @@ def clean_backups(args, log):
-	        to_delete.remove(key_to_keep)
+            keep_file = True # assume we are keeping the file - this is the
+            # default behavior
+            # if we are checking if we should keep existing files
```
Did the `2to3` tool split this comment over two lines?

kimlew replied:
Maybe PyCharm or Atom did. Moved & combined 2 comment lines into 1.

kcharter replied:
OK, confirmed conversion to just one comment on the previous line. Looks good.

kcharter commented on tablechop:
```diff
@@ -75,25 +75,27 @@ def clean_backups(args, log):
+
+            keep_jkey = keep_jkey or keep_file  # keep the jkey if we are
+            # keeping any of the files
+            key_to_keep = '%s:%s' % (args.name, x)
```
Looks like another comment split over two lines here.

kimlew replied:
Also moved & combined 2 comment lines into 1.

kcharter replied:
And confirmed conversion to a single comment on the previous line. Also looks good.

kcharter commented on test_tablesnap.sh:
```diff
@@ -0,0 +1,114 @@
+export AWS_ACCESS_KEY_ID="$AWS_ACC_KEY_ID"
+export AWS_SECRET_ACCESS_KEY="$AWS_SEC_ACC_KEY"
+
+aws s3 ls
```
Is there a reason to run `aws s3 ls` here?

kimlew replied:
Removed this unneeded debugging line.

kcharter replied:
Confirmed removed in an additional commit.

kcharter reviewed (commented):
I have made one pass through this and have a comment and a couple of questions. I will have some more comments on the `Makefile` but I'd like to play with it a bit.

kimlew committed:
Moves 2 comments that were split so each are on 1 line

kimlew committed:
Removes unneeded debugging line

kcharter commented on Makefile:
```diff
@@ -14,7 +14,55 @@ build:
+TIMEOUT_B=10
+AWS_ACC_KEY_ID := $(shell aws configure get aws_access_key_id)
+AWS_SEC_ACC_KEY := $(shell aws configure get aws_secret_access_key)
+BUCKET=klew-tablesnap-pymigration
```
Oh, the default bucket name probably shouldn't have 'klew' in it. That raises a bit of a problem because for each person who runs this script they will need to have their own bucket and the default should reflect that.

I suggest replacing

```
BUCKET=klew-tablesnap-pymigration
```

with

```
BUCKET=$(USER)-tablesnap-tests
```

For you `$(USER)`. will become `klew` on your machine. I'm suggesting `tests` instead of `pymigration` because this VM could be useful for testing lots of different things.

kimlew replied:
Made change.

kcharter commented on Makefile:
```diff
@@ -14,7 +14,55 @@ build:
+BUCKET=klew-tablesnap-pymigration
+HARDLINKS_PATH=/tmp/tablesnap-hardlinks
+
+docker-test-tablesnap: docker-build
```
This `docker-test-tablesnap` target really needs not just `docker-build` but either the `docker-py27env` target or the `docker-py36env` target, since you need at least one of the installed Python environments in order to run the test script.

However, since we've got particular versions of Python that won't really change often, I suggest you create the Python 2.7 and Python 3.6 environments and install all of the required Python packages as part of `docker-build`. That way you can get rid of the `docker-py27env` and `docker-py36env` targets up above. And you don't need to change the dependencies for `docker-test-tablesnap`.

And despite having talked you into using the "bind" mount before, I think we might simplify things if we get rid of it.

You can use the `COPY` command in your `Dockerfile` to copy `tablesnap` and its companion programs to where you need them. This will get rid of the `py27env` and `py36env` directories you had to put in `.gitignore`, and also get rid of the long `--mount` option you have to give to each of the `docker run` commands.

So here's what you'd need to do in your `Dockerfile`.

After the `WORKDIR` command at the bottom, add the commands

```
RUN virtualenv py27env -p /usr/bin/python
RUN virtualenv py36env -p /usr/bin/python3

COPY requirements.txt .

RUN py27env/bin/pip2 install -r requirements.txt
RUN py36env/bin/pip3 install -r requirements.txt

COPY tablechop tablesnap ...(other programs) .
COPY test_tablesnap.sh .
```

Then you should be able to remove the `docker-py27env` and `docker-py36env` targets entirely, and also remove the long `--mount` option

```
--mount type=bind,source=$(PWD),target=/tablesnap
```

from all the `docker run` commands.

kcharter replied:
Once you have done this, there are a couple of other things to tweak but I think it will be easier to explain if you do this change first.

kcharter replied:
OK, your changes addressed all of this comment. I think that's much simpler and I was able to run the `docker-tablesnap-test`.

kcharter reviewed (changes requested):
I have made a couple of suggestions, one of them small and one of them a little bigger, but I don't think they will be hard to do.

kimlew committed:
Removes old bind mount lines & changes bucket to use $USER & tests

kimlew committed:
Creates VMs with Pythons, installs requirements & dependencies, copies files

kimlew committed:
Edits comments to cite $USER & test for bucket name

kcharter commented on Makefile:
```diff
@@ -22,41 +22,33 @@ docker-build:
+	docker run -it migrate2to3:v1 ./test_tablesnap.sh $(DIRECTORY) $(TIMEOUT_A) $(TIMEOUT_B) $(AWS_ACC_KEY_ID) $(AWS_SEC_ACC_KEY) $(BUCKET) $(HARDLINKS_PATH)
 
 clean:
 	rm -rf py27env
```
Now that you create the `py27env` and `py36env` in the Docker image itself, you don't need to delete them anymore as part of the `clean` target. You can remove the two lines that delete them.

kimlew replied:
Deleted.

kcharter commented on Makefile:
```diff
@@ -22,41 +22,33 @@ docker-build:
+	docker run -it -e USER=root migrate2to3:v1 ./py36env/bin/python ./tableslurp --help
+	# docker run -it migrate2to3:v1 ./py36env/bin/python ./tablechop --help
+	# docker run -it migrate2to3:v1 ./py36env/bin/python ./tablesnap-sunset --help
+	# docker run -it migrate2to3:v1 ./py36env/bin/python ./tablesnap-sync-manifest --help
```
Instead of commenting out the other `--help` test lines, you can make a separate target for each one. This will make the `Makefile` a little longer but at least you can run all the tests. For example:

```
test-help-tablesnap-py36: docker-build
    docker run migrate2to3:v1 ./py36env/bin/pythong ./tablesnap --help

test-help-tableslurp-py36: docker-build
    docker run migrate2to3:v1 -e USER=root migrate2to3:v1 ./py36env/bin/python ./tableslurp --help
```

and so on for the others. BTW, I think we don't need the `-it` options for these `docker run` commands.

kcharter replied:
The new targets look good and I have run all of them without trouble.

kcharter reviewed (commented):
I have a couple of little suggestions that should be be pretty quick, and then perhaps we can write a test for `tableslurp` that downloads a file uploaded by `tablesnap`.

kimlew committed:
Deletes unneeded rm lines since Dockerfile creates the 2 directories

- so if clean up of 2 directories wanted, just delete Docker container

kimlew committed:
Creates separate tests for each migrated file

kimlew committed:
Adds a rule to prune dangling Docker images

kcharter commented on test_tablesnap.sh:
```diff
@@ -0,0 +1,112 @@
+# Then, the script checks if tablesnap noticed the events, e.g., uploaded the
+# file with a command like: aws s3 ls
+
+# TODO: Use Python to run tablesnap.
```
Looks like you can remove this TODO comment now.

kimlew replied:
Removed this and other TODOs & updated comments.

kcharter commented on Makefile:
```diff
@@ -14,7 +14,67 @@ build:
+BUCKET=$(USER)-tablesnap-tests
+HARDLINKS_PATH=/tmp/tablesnap-hardlinks
+
+docker-test-tablesnap: docker-build
```
I suggest renaming the `docker-test-tablesnap` target so it's more consistent with the other test targets above. Maybe `test-tablesnap-upload`?

kimlew replied:
Made 2 rules to replace 1 old rule as we talked about and renamed to:
test-tablesnap-upload-py2.7
test-tablesnap-upload-py3.6

kcharter replied:
New targets added, but I suggested changing their names in the comments below.

kcharter commented on Makefile:
```diff
@@ -14,7 +14,67 @@ build:
+HARDLINKS_PATH=/tmp/tablesnap-hardlinks
+
+docker-test-tablesnap: docker-build
+	docker run -it migrate2to3:v1 ./test_tablesnap.sh $(DIRECTORY) $(TIMEOUT_A) $(TIMEOUT_B) $(AWS_ACC_KEY_ID) $(AWS_SEC_ACC_KEY) $(BUCKET) $(HARDLINKS_PATH)
```
For all the other tests you have a Python 2 version of the test, and a Python 3 version of the test, but for this test you use only Python 3. I suggest adding another test that runs the `test_tablesnap.sh` script but that uses Python 2. You should be able to do this by adding another parameter to the `test_tablesnap.sh` script that tells it whether to use Python 2 or Python 3. By default it could use Python 3.

In the `test_tablesnap.sh` script, you'd use that parameter to decide whether to use `py27env/bin/python` (Python 2) or `py36env/bin/python` to run the `tablesnap` program. If you'd like, we can discuss it looking at the files.

kimlew replied:
Made changes for parameter of the Python version passed in from the Makefile in test_tablesnap.sh & Makefile.

kcharter reviewed (commented):
This looks good -- I've got one trivial comment and one that is a little harder.

I was going to ask you to add a test that checks that we can use the `tableslurp` program to download files, but it turns out that test is a little awkward to write. Unfortunately, the way we are using `tableslurp` it requires files that we upload into a second S3 bucket, and those files we create using a script that lives in another repository that isn't public.

I will give this stuff a try in one of our test environments and if it all works OK I'll give this the official approval.

kimlew committed:
Changes rule name so more consistent with other rule names

kimlew committed:
Updates comments & the check for number of args passed

kimlew committed:
Separates test into 2 tests for upload based on Python version used

kimlew committed:
Adds passed param of Python version & check for this, updates comments

kcharter commented on Makefile:
```diff
@@ -66,8 +66,11 @@ AWS_SEC_ACC_KEY := $(shell aws configure get aws_secret_access_key)
+	docker run -it migrate2to3:v1 ./test_tablesnap.sh python2 $(DIRECTORY) $(TIMEOUT_A) $(TIMEOUT_B) $(AWS_ACC_KEY_ID) $(AWS_SEC_ACC_KEY) $(BUCKET) $(HARDLINKS_PATH)
+
+test-tablesnap-upload-py3.6: docker-build
+	docker run -it migrate2to3:v1 ./test_tablesnap.sh python3 $(PYTHON2) $(DIRECTORY) $(TIMEOUT_A) $(TIMEOUT_B) $(AWS_ACC_KEY_ID) $(AWS_SEC_ACC_KEY) $(BUCKET) $(HARDLINKS_PATH)
```
Oh, I think you need to remove the `$(PYTHON2)` argument right after the `python3`. It turns out that doesn't hurt anything because it expands to an empty string, but it's confusing to read.

kimlew replied:
Good catch. It's true, when I ran the test, there was no error. Removed $(PYTHON2) & re-tested rule.

kcharter reviewed (changes requested):
Looks good, I found one little mistake.

kimlew committed:
Removed $(PYTHON2) from rule, test-tablesnap-upload-py3.6

kcharter reviewed (commented):
Latest change looks good @kimlew. I think adding a test for `tableslurp` downloads is about all we need now.

kimlew committed:
Corrects comments

kimlew committed:
Adds test for tableslurp starting with test-tableslurp-upload rules

kimlew committed:
Adds retrieval/download of test file from S3

- Renames local test file - to see difference between original & downloaded file
- Downloads file from S3 & checks for downloaded file

Note: There is an error with tableslurp test for Python 3.6:
...
  File "./tableslurp", line 140, in _build_file_set
    self.fileset = set(filter(bool, key.get_contents_as_string().split('\n')))
TypeError: a bytes-like object is required, not 'str'

kimlew committed:
Fixes Python 3.6 test error. Adds arg, encoding='utf-8'

Note: get_contents_as_string() needs encoding argument
- encoding defaults to return bytes, so you must turn arg type into a string

See: http://boto.cloudhackers.com/en/latest/ref/s3.html
- get_contents_as_string() section
- encoding sub-section

kcharter commented on Makefile:
```diff
@@ -72,6 +72,12 @@ test-tablesnap-upload-py2.7: docker-build
 test-tablesnap-upload-py3.6: docker-build
 	docker run -it migrate2to3:v1 ./test_tablesnap.sh python3 $(DIRECTORY) $(TIMEOUT_A) $(TIMEOUT_B) $(AWS_ACC_KEY_ID) $(AWS_SEC_ACC_KEY) $(BUCKET) $(HARDLINKS_PATH)
 
+test-tableslurp-upload-py2.7: docker-build
```
Trivial: this target should be called `test-tableslurp-download-py2.7`, since `tableslurp` downloads files from S3. Similarly with the `test-tableslurp-upload-py3.6` target.

kimlew replied:
Right, it is a download. Changed.

kcharter replied:
Targets renamed as requested.

kcharter commented on test_tableslurp.sh:
```diff
@@ -0,0 +1,142 @@
+echo "INDEX_DEST is:" "${INDEX_DEST}"
+echo
+
+touch "${SRC_LOCAL}" > /dev/null
```
Suggestion: instead of using `touch` to create the `$SRC_LOCAL` file, redirect the output of `date` into the file:

```
date > "${SRC_LOCAL}"
```

This would allow us to make a more interesting test later on, where we can compare the content of the file that `tableslurp` downloads to this file (after it's renamed)

kimlew replied:
Good suggestion. Changed.

kcharter replied:
Confirmed using `date` to give the file some contents.

kcharter commented on test_tableslurp.sh:
```diff
@@ -0,0 +1,142 @@
+echo
+
+touch "${SRC_LOCAL}" > /dev/null
+touch "${INDEX_SRC}" > /dev/null
```
You don't need to `touch "${INDEX_SRC}"` because you redirect `echo` to it below.

kimlew replied:
Oh right. This line will create it:
echo "${DIR_LOCAL}/${TEST_FILE}" > "${INDEX_SRC}"
Deleted touch line.

kcharter replied:
Confirmed you removed the unnecessary line.

kcharter commented on test_tableslurp.sh:
```diff
@@ -0,0 +1,142 @@
+# tableslurp & check if file exists to prove file was retrieved from AWS.
+echo "${DIR_LOCAL}/${TEST_FILE}"
+echo "${DIR_LOCAL}/$RENAMED_FILE"
+mv "${DIR_LOCAL}/${TEST_FILE}" "${DIR_LOCAL}/${RENAMED_FILE}"
```
You want to rename the file _before_ you run `tableslurp`. Otherwise `tableslurp` may overwrite the file before you get a chance to `mv` it.

Oh, I see, you are running `tableslurp` twice. You don't need to do that. The steps you need are

1. use `aws s3 cp` to copy the test file and the index file into S3
2. use `mv` to rename the test file
3. run `tableslurp` to download the test file from S3
4. (suggestion): use `diff` to check whether the renamed test file from step 2 and the one downloaded by `tableslurp` in step 3 have the same contents

kimlew replied:
Yes, the first time I ran `tableslurp` was just to confirm it worked before and after the mv.
Deleted first run.

Good suggestion with the diff check. Changed.

kcharter replied:
Confirmed script changed as described.

kcharter commented on test_tableslurp.sh:
```diff
@@ -0,0 +1,142 @@
+echo "${DIR_LOCAL}/$RENAMED_FILE"
+mv "${DIR_LOCAL}/${TEST_FILE}" "${DIR_LOCAL}/${RENAMED_FILE}"
+
+# Check if test file is temporarily removed from /tmp/tablesnap-test.
+if [ -f "${DIR_LOCAL}/${TEST_FILE}" ]; then
+  echo "${TEST_FILE} was NOT deleted. Directory has:"
+  ls "${DIR_LOCAL}"
+else
+  echo "${TEST_FILE} has been deleted. Directory now has:"
+  ls "${DIR_LOCAL}"
+fi
+echo
```
Suggestion: I don't think you really need this block. Are you trying to check whether the `mv` command above worked?

kimlew replied:
Yes, it's just a check that the mv occurred. I will remove for the final version.

kcharter replied:
Okie dokie.

kcharter commented on test_tableslurp.sh:
```diff
@@ -0,0 +1,142 @@
+
+./${PYTHON_ENV}/bin/python ./tableslurp -k "$AWS_ACC_KEY_ID" -s "$AWS_SEC_ACC_KEY" --name "${CONTAINER}" "${INDEX_BUCKET}" "${INDEX_FILE}" "${BACKUP_BUCKET}"
+echo
+echo "Re-running tableslurp."
```
This should be the only time you need to run `tableslurp`

kimlew replied:
Right. As I mentioned in a few comments above, the first tableslurp run is deleted.

kcharter replied:
Confirmed you're running `tableslurp` just once now.

kcharter commented on test_tableslurp.sh:
```diff
@@ -0,0 +1,142 @@
+echo "Re-running tableslurp."
+# Check if successful download, i.e., if file exists locally in /tmp/tablesnap-test.
+if [ -f "${DIR_LOCAL}/${TEST_FILE}" ]; then
+  echo "${TEST_FILE} was retrieved. Directory has:"
+  ls "${DIR_LOCAL}"
```
Once you know `"${DIR_LOCAL}/${TEST_FILE}"` exists, I suggest using `diff` to check that it has the expected contents. `diff` will exit with status zero if two files are the same, so you can use this:

```
if ! diff "${DIR_LOCAL}/${TEST_FILE}" "${DIR_LOCAL}/${RENAMED_FILE}"; then
  echo "${TEST_FILE} and ${RENAMED_FILE} have different contents"
  exit 1
fi
```

kimlew replied:
Changed.

kcharter replied:
Looks good to me.

kcharter commented on test_tableslurp.sh:
```diff
@@ -0,0 +1,142 @@
+else
+  echo "${TEST_FILE} was NOT retrieved. Directory has:"
+  ls "${DIR_LOCAL}"
+fi
```
In this case, `tableslurp` has not downloaded the file, and that is really an error. Whenever something doesn't happen the way it should, your script should exit with a non-zero status code to indicate that something went wrong. I think it's fine to use `exit 1` to exit with status code one.

In my own scripts I'm fond of defining a function

```
die() {
  >&2 echo "error: $*"
  exit 1
}
```

so that you can write things like

```
die "tableslurp did not download the file"
```

to exit with a message and a non-zero status code.

kimlew replied:
I have added this function and the function call to the last check for the downloaded file.

Note: I have left all debugging echo statements in the code for now. I will remove them once the test doesn't need any more changes.

kcharter replied:
Looks good.

kcharter reviewed (changes requested):
Looking good, but there are a few things to fix and I've got a couple of suggestions.

kimlew committed:
Renames target rule to download

kimlew committed:
Changes last download check-if error, exit. Uses date as file contents, etc.

kimlew committed:
Adds die() function to exit script when download fails

kcharter commented on test_tableslurp.sh:
```diff
@@ -50,32 +55,32 @@ export AWS_SECRET_ACCESS_KEY="$AWS_SEC_ACC_KEY"
-echo "AFTER touch & > index file, the ls of /tmp/tablesnap-test:"
+date > "${SRC_LOCAL}"
+echo "${SRC_LOCAL}" > "${INDEX_SRC}"
+echo "CHECK /tmp/tablesnap-test with ls after ${SRC_LOCAL} given content & $UPDOWN_FILE > index file:"
```
I think you meant to use `${UPDOWN_FILE}` with curly braces, just to be consistent.

kimlew replied:
It's just a debugging statement but I've reworded it so it's more clear.

kcharter reviewed (commented):
I've got just one very small comment now. I'll have a look again after I have tried the `tableslurp-download` tests myself.

kcharter reviewed (commented):
When I run `make test-tableslurp-download-py2.7`  it succeeds but prints this message:

```
FAILURE: INDEX_FILE has NOT been copied to S3.
```

The same thing happens when I run `make test-tableslurp-download-py3.6`.

That is probably just a bug in your script.

kimlew committed:
Changes index.txt on S3 check to use container, tester, in path & renames vars

kcharter commented on test_tableslurp.sh:
```diff
@@ -109,44 +110,50 @@ echo
 else
   echo "${UPDOWN_FILE} has been deleted. Directory now has:"
-  ls "${DIR_LOCAL}"
+  ls -lah "${DIR_LOCAL}"
```
I suggest calling `die` with a message here. If you can't `mv` the file, there is something wrong with the test.

kcharter replied:
Looks good.

kcharter commented on test_tableslurp.sh:
```diff
@@ -59,45 +59,46 @@ UPDOWN_FILE="aws_s3_cp.txt"
+if aws s3 ls "${DEST_S3_INDEX_FILE}" | grep "$INDEX_FILE"; then
 	echo "SUCCESS: INDEX_FILE has been copied to S3."
 else
 	echo "FAILURE: INDEX_FILE has NOT been copied to S3."
```
I suggest calling `die` with a message here. If you can't upload the index file to S3, something has gone wrong and you shouldn't continue the test.

kimlew replied:
Done.

kcharter replied:
Looks good.

kcharter commented on test_tableslurp.sh:
```diff
@@ -59,45 +59,46 @@ UPDOWN_FILE="aws_s3_cp.txt"
 	echo "SUCCESS: UPDOWN_FILE has been copied to S3."
 else
 	echo "FAILURE: UPDOWN_FILE has NOT been copied to S3."
 fi
```
I suggest calling `die` with a message here. If you can't copy the test file to S3, you shouldn't continue the test.

kimlew replied:
Done, plus added a couple more calls to die function & added some string constants so code lines a bit shorter.

kcharter replied:
Looks good.

kcharter reviewed (commented):
I think renaming the variables has improved things, making it easier to see what's going on.

There are a few more places where I think you should call `die` with a message -- I have marked them with separate comments.

kimlew committed:
Adds calls to die function & string constants so code lines shorter

kcharter commented on test_tableslurp.sh:
```diff
@@ -71,87 +69,84 @@ BACKUP_BUCKET="${BUCKET}"
+# Check if test_aws_s3_cp.txt was successfully downloaded locally to
+# /tmp/tablesnap-test. Note: if ! diff - means there IS a difference between
+# the 2 files, i.e., returns a status code of 1, & there IS an ERROR. If
+# ! diff returns a status code of 0, there is NO difference between the files,
```
Did you mean `If diff returns a status code of 0, there is NO difference...`?

kimlew replied:
I found it takes too much time to remember what diff actually returns, which requires you to remember it returns a status code of 0 or 1 and then you have to figure out the negation part. 
As an aside, 'cmp' acts the same way as 'diff' so these 2 lines return the same results, e.g., returns status code 1 so is an ERROR:

'''if ! diff "${SRC_LOCAL_UPDOWN_FILE}" "${SRC_LOCAL_ORIG_TEST_FILE}"; then'''
'''if ! cmp --silent "${SRC_LOCAL_UPDOWN_FILE}" "${SRC_LOCAL_ORIG_TEST_FILE}"; then'''

As you suggested, I used a case statement instead. I think this makes it faster to see that diff returns a status code & what the status code means.

kcharter replied:
Case statement looks good.

kcharter commented on test_tableslurp.sh:
```diff
@@ -71,87 +69,84 @@ BACKUP_BUCKET="${BUCKET}"
-# downloaded files.
+echo "Running tableslurp..."
+echo
+
```
Since `tableslurp` does its thing and exits, I suggest moving the `echo "Running tableslurp..."` to just before the `./${PYTHON_ENV}/bin/python ./tableslurp...` line. Otherwise, you're printing that message after `tableslurp` has already run.

kimlew replied:
Oh, that order is better.

kcharter replied:
Confirmed moved.

kcharter reviewed (commented):
This is looking pretty good, and you found some more places to call `die` that I hadn't noticed!

Just a couple of minor things I noticed: possible error in a comment, and a suggestion to move one of your `echo` commands.

kimlew committed:
Changes last check to a "case" & some echo statements to printf

case statement makes it faster to see that diff returns a status code & what the status code means

kcharter commented on test_tableslurp.sh:
```diff
@@ -128,27 +128,28 @@ echo
+    echo "Orig file contents: "
+    cat "${SRC_LOCAL_ORIG_TEST_FILE}"
+    die "${FAILURE} download ${UPDOWN_FILE} from S3"
+    ;;
```
`diff` can very likely return other non-zero status codes to indicate various errors. It's a good idea to add a catch-all at the end, just before the `esac` where you exit with a message like

```
  *)
    die "diff returned status ${?}"
    ;;
```

kimlew replied:
Whoops. I forgot about the catch-all. Thanks.

kcharter replied:
Looks good.

kcharter commented on test_tableslurp.sh:
```diff
@@ -71,10 +71,10 @@ BACKUP_BUCKET="${BUCKET}"
-echo "Dest on S3 for file to upload:\t" "${DEST_S3_UPDOWN_FILE}"
-echo "Source on local for index file to upload:" "${SRC_LOCAL_INDEX_FILE}"
-echo "Dest on S3 for index file to upload:\t" "${DEST_S3_INDEX_FILE}"
+printf "%s\n Source on local for file to upload:" "${SRC_LOCAL_UPDOWN_FILE}"
+printf "%s\n Dest on S3 for file to upload:\t" "${DEST_S3_UPDOWN_FILE}"
+printf "%s\n Source on local for index file to upload:" "${SRC_LOCAL_INDEX_FILE}"
+printf "%s\n Dest on S3 for index file to upload:\t" "${DEST_S3_INDEX_FILE}"
```
I think you meant to put the `%s \n` at the end of the first strings, not the beginning. For example

```
printf "Source on local for file to upload: %s\n" "${SRC_LOCAL_UPDOWN_FILE}"
```

instead of

```
printf "%s\n Source on local for file to upload:" "${SRC_LOCAL_UPDOWN_FILE}"
```

Right now you're getting funny output that looks like this:

```
/tmp/tablesnap-test/aws_s3_cp.txt
 SRC_LOCAL_UPDOWN_FILE:		/tmp/tablesnap-test/orig_test_file.txt
 SRC_LOCAL_ORIG_TEST_FILE:
```

kimlew replied:
I missed seeing that in the output. Thanks.

kcharter replied:
Confirmed fixed.

kcharter commented on Makefile:
```diff
@@ -14,7 +14,76 @@ build:
+BUCKET=$(USER)-tablesnap-tests
+HARDLINKS_PATH=/tmp/tablesnap-hardlinks
+
+test-tablesnap-upload-py2.7: docker-build
+	docker run -it migrate2to3:v1 ./test_tablesnap.sh python2 $(DIRECTORY) $(TIMEOUT_A) $(TIMEOUT_B) $(AWS_ACC_KEY_ID) $(AWS_SEC_ACC_KEY) $(BUCKET) $(HARDLINKS_PATH)
+
+test-tablesnap-upload-py3.6: docker-build
+	docker run -it migrate2to3:v1 ./test_tablesnap.sh python3 $(DIRECTORY) $(TIMEOUT_A) $(TIMEOUT_B) $(AWS_ACC_KEY_ID) $(AWS_SEC_ACC_KEY) $(BUCKET) $(HARDLINKS_PATH)
+
+test-tableslurp-download-py2.7: docker-build
+	docker run -e USER=root -it migrate2to3:v1 ./test_tableslurp.sh python2 $(DIRECTORY) $(AWS_ACC_KEY_ID) $(AWS_SEC_ACC_KEY) $(BUCKET)
+
+test-tableslurp-download-py3.6: docker-build
+	docker run -e USER=root -it migrate2to3:v1 ./test_tableslurp.sh python3 $(DIRECTORY) $(AWS_ACC_KEY_ID) $(AWS_SEC_ACC_KEY) $(BUCKET)
```
You should remove the `-it` on the `docker run` command lines in these four test targets too, just like you did with the `test-help-...` targets above. The `-it` isn't necessary and it will cause failures if you try to run `make` when standard input isn't a terminal.

For example, I tried to run all your targets that are named `test-...` by doing this:

```
grep '^test' Makefile | cut -f 1 -d: | xargs -n1 make
```

The `test-help-...` targets all work fine, but the `test-tablesnap-upload-...` and `test-tableslurp-download-...` targets fail because `docker run` complains that its standard input isn't a terminal.

kimlew replied:
Deleted -it from lines that don't need user interaction.

kcharter replied:
Looks good and I confirmed that I can now run all the tests with

```
grep '^test' Makefile | cut -f 1 -d: | xargs -n1 make
```

kcharter reviewed (commented):
I agree that the `case` statement is definitely clearer. I have a suggestion for that, and pointed out a problem with your `printf` commands. I also suggest removing the `-it` options for `docker run` from your last four tests.

kimlew committed:
Deleted -it from lines that do not need user interaction & TTY

kimlew committed:
Adds catch-all case & fixes printf placeholder position

kcharter reviewed (commented):
This looks good and the tests seem to work smoothly. I've got a couple more little suggestions for tweaks to the makefile.

First, it would be nice to have a target for removing all the files in the test bucket. You can do this with an `aws` CLI command like this:

```
aws s3 rm --recursive s3://$(BUCKET)/
```

This doesn't destroy the test bucket, it just removes all its contents. If you add a new target in the `Makefile`, say `empty-test-bucket`, that uses that `aws s3 rm` command above, then we could empty out the test bucket with

```
make empty-test-bucket
```

My second suggestion is to add a `test` target that depends on all your specific tests, so that you can run all tests with

```
make test
```

To do this you add lines declaring that `test` depends on your other tests:

```
test: test-help-tablesnap-py2.7
test: test-help-tableslurp-py2.7
...
```

one line for each of your other tests.

You should declare the `test` target and the `empty-test-bucket` targets to be "phony" targets, with

```
.PHONY: empty-test-bucket
.PHONY: test
``` 

These are just suggestions, but I think they'd be nice additions.

Either way, I think we're ready to give this a whirl!

kimlew commented:
Should I also add this line?

`.PHONY: clean-docker`

kimlew committed:
Adds target rule, test - so easier to run. Adds 2 .PHONY rules.

kcharter commented:
> Should I also add this line?
> 
> `.PHONY: clean-docker`

If it's not already there, yes. Any target that isn't the name of a file should really be declared 'phony'.

kcharter commented:
@kimlew I have tried out both `make empty-test-bucket` and `make test` and they work as I expected. The nice thing about `make test` is it rebuilds the Docker image only once at the beginning, and so runs through all the tests faster than doing them one at a time or using that `xargs` command pipeline.

kimlew committed:
Add other missed .PHONY target rules

kimlew committed:
Increases timeout to remove race condition

kimlew committed:
Adds some vertical whitespace so easier to read output

kimlew commented:
Re: `test_tableslurp.sh` - I added more target rules to the .PHONY target rule.

Re: `test_tablesnap.sh` - I increased the timeout to run tablesnap from 10 to 15 seconds to see if that eliminates the race condition you found (which never appeared for me). And I added a bit of whitespace to make the output easier to read.

kcharter reviewed (commented):
Additional changes to add `.PHONY` target declarations, increase your second timeout for the `tablesnap` upload tests, and tweak the spacing for output all look good to me.

Before I formally approve I'll try this out in a test environment, just in case that catches some other things that need to change. I should be able to get that going in the next few days.

kcharter commented:
@kimlew ah it turns out that there are three other scripts that call `get_contents_as_string()` on an S3 key without passing `encoding='utf-8'`:

- once in `tablechop`
- once in `tablesnap-backfill-size`
- once in `tablesnap-sync-manifest`

It turns out we're definitely using the last one, but they all need to be fixed the same way you did in `tableslurp`.

dfriel commented on test_tableslurp.sh:
```diff
@@ -0,0 +1,159 @@
+#! /bin/sh
+
+# Script Name: test_tableslurp.sh
```
Suggestion: add a description and a usage() function like test_tablesnap.sh has.

kimlew replied:
Added.

dfriel commented:
Makefile targets all work and shellcheck passes for test_tableslurp.sh and test_table_snap.sh.

dfriel commented:
`tablesnap-backfill-size` has a reference to `iteritems` and `xrange`, but I suspect we don't use that script anywhere.

kimlew committed:
Encodes string to utf-8 for key

kimlew committed:
Changes tablesnap-backfill-size for Python migration

kimlew committed:
Adds purpose & usage, fixes some spacing & indentation

kimlew commented:
> @kimlew ah it turns out that there are three other scripts that call `get_contents_as_string()` on an S3 key without passing `encoding='utf-8'`:
> 
> * once in `tablechop`
> * once in `tablesnap-backfill-size`
> * once in `tablesnap-sync-manifest`
> 
> It turns out we're definitely using the last one, but they all need to be fixed the same way you did in `tableslurp`.

Added the encoding.

kimlew commented:
> `tablesnap-backfill-size` has a reference to `iteritems` and `xrange`, but I suspect we don't use that script anywhere.

The 2to3 tool changed these to items() and range(). Good catch. I changed them just in case `tablesnap-backfill-size` is used.

kcharter commented on test_tableslurp.sh:
```diff
@@ -2,8 +2,30 @@
 # Script Name: test_tableslurp.sh
 
+# Purpose: This script is called by the Makefile which passes several arguments
+# and then runs tableslurp in the background, e.g., ./tableslurp& to download a
```
Ah, I didn't see it yesterday, but you actually run `tableslurp` in the foreground; `tableslurp` isn't a service like `tablesnap`, but a program that you run when you need it. So you could just say

```
... then runs tableslurp to download a file ...
```

kimlew replied:
Oh right. Changed.

kcharter commented on test_tableslurp.sh:
```diff
@@ -2,8 +2,30 @@
+usage() {
+	cat <<-EOF
+	USAGE: $(basename "$0") DIRECTORY AWS_ACC_KEY_ID AWS_SEC_ACC_KEY BUCKET
+	Downloads a file from the target directory, /tmp/tablesnap-test.
```
Downloads a file from S3 to the test directory?

kimlew replied:
Missed that. Thanks. Changed.

kcharter reviewed (commented):
Looks good, just a couple of minor suggestions about the purpose and usage stuff you added.

I re-ran `make test` a couple of times and it succeeded both times.

kimlew committed:
Changes incorrect description

kimlew committed:
Changes order of words & uses test vs. target for directory

kcharter reviewed (approved):
Looks good to me.

kcharter commented:
@kimlew @dfriel and I think your changes look good and we've been testing them over the last five days without trouble. I'm going to merge your PR. Thanks for the contribution!
